src/tamp_improv/approaches/improvisational/training.py

```diff
@@ -0,0 +1,242 @@
+
+    # Visualization settings
+    render: bool = False
+    record_training: bool = False
+    training_record_interval: int = 50
+
+    # Policy-specific settings
+    policy_config: dict[str, Any] | None = None
+
+    def get_training_data_path(self, system_name: str) -> Path:
+        """Get path for training data for specific system."""
+        return Path(self.training_data_dir) / system_name
+
+
+@dataclass
+class Metrics:
+    """Training and evaluation metrics."""
+
+    success_rate: float
+    avg_episode_length: float
+    avg_reward: float
+
+
+def get_or_collect_training_data(
+    system: ImprovisationalTAMPSystem[ObsType, ActType],
+    approach: ImprovisationalTAMPApproach[ObsType, ActType],
+    config: TrainingConfig,
+) -> TrainingData:
+    """Get existing or collect new training data."""
+    # Check if saved data exists
+    data_path = config.get_training_data_path(system.name)
+
+    if not config.force_collect and data_path.exists():
+        print(f"\nLoading existing training data from {data_path}")
+        try:
+            train_data = TrainingData.load(data_path)
+
+            # Verify config matches
+            if (
+                train_data.config.get("seed") == config.seed
+                and train_data.config.get("collect_episodes") == config.collect_episodes
+                and train_data.config.get("max_steps") == config.max_steps
+            ):
+                print(f"Loaded {len(train_data)} training scenarios")
+                return train_data
+            print("Existing data has different config, collecting new data...")
+        except Exception as e:
+            print(f"Error loading training data: {e}")
+            print("Collecting new data instead...")
+
+    # Collect new data
+    train_data = collect_training_data(system, approach, config)
+
+    # Save the collected data
+    print(f"\nSaving training data to {data_path}")
+    train_data.save(data_path)
+
+    return train_data
+
+
+def collect_training_data(
+    system: ImprovisationalTAMPSystem[ObsType, ActType],
+    approach: ImprovisationalTAMPApproach[ObsType, ActType],
+    config: TrainingConfig,
+) -> TrainingData:
+    """Collect training data from TAMP execution."""
+    training_states: list[Union[int, ObsType]] = []
+    training_preconditions = []
+
+    print("\nCollecting training data...")
+
+    # Run episodes to collect states where preconditions aren't met
+    for episode in range(config.collect_episodes):
+        print(f"\nCollection episode {episode + 1}/{config.collect_episodes}")
+
+        obs, info = system.reset()
+        approach.reset(obs, info)
+
+        for _ in range(config.max_steps):
+            # If we encounter a state where preconditions aren't met
+            if approach.policy_active:
+                if isinstance(obs, int):
+                    training_states.append(obs)
+                else:
+                    training_states.append(deepcopy(obs))
```

Comment:
why is deepcopy needed here? (it could be slow if the observations get complex)

Reply:
First, the reason why we're using `copy()` at all is because appending `obs` directly to `training_states` caused unintended changes due to mutable references. Second, we couldn't use `copy()` directly because `ObsType` is a generic type parameter and Python's type checker doesn't know if it has a copy method (I encountered the error running ci checks). That's why the choice landed on deepcopy. Let me know if you know any better way to do this! I totally agree that this could make the process slower in the future.

Reply:
> appending obs directly to training_states caused unintended changes due to mutable references.

can you elaborate on this part? that might be indicative of some other issues. why would training modify states?

Reply:
I tried a couple different settings and I realized that the only thing we need to do copy for is
```
# Get positions from current state
            reset_options = {
                "robot_pos": current_state[0:2].copy(),
                "block_1_pos": current_state[4:6].copy(),
                "block_2_pos": current_state[6:8].copy(),
            }
```
in `reset()` of `blocks2d_wrappers.py`. Without `copy()`, `current_state[0:2]` creates a view into the original array that was stored in `training_states`. And when the environment uses these positions to set its internal state and modifies `self.robot_position`, it'd be modifying the same memory that our stored training state points to as well.

But yes we don't need this deepcopy() in `training.py`. Thanks for checking!